File: run_lhe.py
import sys, os
import matplotlib.pyplot as plt
import math

print(os.getcwd())
cfg = open("pythia8310/Makefile.inc")
lib = "../lib"
for line in cfg:
    if line.startswith("PREFIX_LIB="): lib = line[11:-1]; break
sys.path.insert(0, lib)
print(lib)
import argparse


import pythia8
import datetime

# Configuration de Pythia
pythia = pythia8.Pythia()
pythia.readFile("pythia_config.cmnd")  # Assurez-vous d'avoir le bon fichier de config
pythia.init()

# Préparer le fichier de sortie LHE
output_filename = "output.lhe"
with open(output_filename, "w") as file:
    now = datetime.datetime.now()
    file.write(f"<!--\n  File written by custom Pythia8 Python script on {now.strftime('%d %b %Y at %H:%M:%S')}\n-->\n")
    file.write("<LesHouchesEvents version=\"1.0\">\n")
    # Insérez ici les données d'initialisation
    file.write("<init>\n")
    file.write(" 2212  2212  6.500000e+03  6.500000e+03  0  0  0  0  3  1\n")
    file.write(" 8.218007e+04  3.901604e+02  1.000000e+00   9999\n")
    file.write("</init>\n")
    
# Nombre total d'événements à générer
    total_events = 10000  # Modifiez selon le nombre d'événements désiré

    for iEvent in range(total_events):
        if not pythia.next():
            continue

        # Commencer le bloc <event>
        file.write("<event>\n")
        # Le nombre de particules dans l'événement, l'ID du processus, etc.
        nParticles = pythia.event.size() - 2  # Exclure les particules du beam
        file.write(f"    {nParticles} 9999 1.000000e+00 WEIGHT SCALE PDF1 PDF2 ...\n")

        for i in range(pythia.event.size()):
            part = pythia.event[i]
            if part.isFinal() or part.idAbs() in [21, 22]:  # Filtrer pour les particules finales, gluons, et photons
                # Écrire les données de la particule
                file.write(f"    {part.id()} {part.status()} 0 0 0 0 {part.px()} {part.py()} {part.pz()} {part.e()} {part.m()} 0. 9.\n")

        file.write("</event>\n")
        
        
        # Finaliser le fichier LHE
    file.write("</LesHouchesEvents>\n")

# Statistiques Pythia
pythia.stat()

File: run_lhe.cpp
#include "Pythia8/Pythia.h"

using namespace Pythia8;

#include <iostream>
#include <fstream>
#include <string>

int main(int argc, char* argv[]) {
    // Vérifier les arguments
    if (argc < 4) {
        std::cerr << "Usage: " << argv[0] << " <commandFile> <outputFile> <nEvents>" << std::endl;
        return 1;
    }
    std::string commandFile = argv[1];
    std::string outputFile = argv[2];
    int nEvents = std::stoi(argv[3]);

    // Créer une instance de Pythia
    Pythia pythia;

    // Lire le fichier de commande
    pythia.readFile(commandFile);

    // Définir le nombre d'événements
    pythia.readString("Main:numberOfEvents = " + std::to_string(nEvents));

    // Initialiser Pythia
    pythia.init();

    // Créer une instance de LHAup pour écrire le fichier LHE
    LHAupFromPYTHIA8 up(pythia.event, pythia.info);
    up.openLHEF(outputFile);

    // Générer les événements
    int nAbort = 0;
    int maxErr = pythia.mode("Main:timesAllowErrors") + 10;
    for (int iEvent = 0; iEvent < nEvents; ++iEvent) {
        if (!pythia.next()) {
            nAbort++;
            if (nAbort >= maxErr) {
                std::cerr << "Event generation ended prematurely at Event " << iEvent
                          << ", due to too many errors (" << maxErr << ")" << std::endl;
                break;
            }
            continue;
        }

        // Écrire l'événement dans le fichier LHE
        up.setEvent();
        up.eventLHEF();

        // Ici, vous pouvez ajouter du code pour traiter chaque événement si nécessaire
    }

    // Imprimer les statistiques des événements générés
    pythia.stat();

    // Fermer le fichier LHE
    up.closeLHEF(true);

    return 0;
}
